### DISCUSSION TITLE
Read members of referenced classes outside solution

### DISCUSSION THREAD

**Discussion by juripyykko** [dotnet/roslyn#61808] [Q&A] [answered]
I'm experimenting with Roslyn, and am trying to do stuff with class members: I find a class, then I go through it's members, and then recursively do the same for the member types.

The approach below works with classes in my own solution. But when the type is "NamedType System.Int32" for example, then the code below fails because DeclaringSyntaxReferences has no elements.

```
var typeSyntaxRef = typeSymbol.DeclaringSyntaxReferences.Single();
var typeSyntax = (ClassDeclarationSyntax)typeSyntaxRef.GetSyntax();

var fieldSyntaxes = typeSyntax
      .DescendantNodes()
      .OfType<FieldDeclarationSyntax>();
```
Is there some way I can access what I can see in VS in the "Int32 [from metadata]" tab below? I would like to get the member types and names at least.
![image](https://user-images.githubusercontent.com/4171882/173060865-4698c71a-524d-49b3-bed1-10bb20c913ae.png)

**Comment by PathogenDavid** [comment, marked as answer]
Symbols from outside your project are metadata-only symbols and don't have syntax since they don't come from a syntax tree.

That view in Visual Studio is just C# generated based on the metadata symbols, which is why it doesn't actually match the real source code. (For example, in [the actual source](https://github.com/dotnet/runtime/blob/e026f392b2bb93bc9509ac5403d3e8cd7d3eee72/src/libraries/System.Private.CoreLib/src/System/Int32.cs#L29-L30), `MaxValue` and `MinValue` are defined using hex literals.)

You need to stick to using the `ISymbol` API. (For example, use `INamespaceOrTypeSymbol.GetMembers` instead of `DescendantNodes`.)

You may find it helpful to implement a [`SymbolVisitor`](https://docs.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.symbolvisitor?view=roslyn-dotnet-4.2.0) to make it easier to process things.

**Reply by juripyykko** [reply]
Great, that worked for what I was trying to achieve. Thanks!
